astroid/constraint/constraint.py

```diff
@@ -0,0 +1,138 @@
+# Licensed under the LGPL: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html
+# For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
+
+"""Classes representing different types of constraints on inference values."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import TypeVar, Union
+
+from astroid import nodes, util
+
+__all__ = ("get_constraints",)
+
+
+NameNodes = Union[nodes.AssignAttr, nodes.Attribute, nodes.AssignName, nodes.Name]
+ConstraintT = TypeVar("ConstraintT", bound="Constraint")
+
+
+class Constraint:
+    """Represents a single constraint on a variable."""
+
+    def __init__(self, node: nodes.NodeNG, negate: bool) -> None:
+        self.node = node
+        """The node that this constraint applies to."""
+        self.negate = negate
+        """True if this constraint is negated. E.g., "is not" instead of "is"."""
+
+    @classmethod
+    @abstractmethod
+    def match(
+        cls: ConstraintT, node: NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> ConstraintT | None:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        If negate is True, negate the constraint.
+        """
+
+    @abstractmethod
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+
+
+class NoneConstraint(Constraint):
+    """Represents an "is None" or "is not None" constraint."""
+
+    CONST_NONE: nodes.Const = nodes.Const(None)
+
+    @classmethod
+    def match(
+        cls: ConstraintT, node: NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> ConstraintT | None:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        Negate the constraint based on the value of negate.
+        """
+        if isinstance(expr, nodes.Compare) and len(expr.ops) == 1:
+            left = expr.left
+            op, right = expr.ops[0]
+            if op in {"is", "is not"} and (
+                matches(left, node)
+                and matches(right, cls.CONST_NONE)
+                or matches(left, cls.CONST_NONE)
+                and matches(right, node)
+            ):
+                negate = (op == "is" and negate) or (op == "is not" and not negate)
+                return cls(node=node, negate=negate)
+
+        return None
+
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+        # Assume true if uninferable
+        if inferred is util.Uninferable:
+            return True
+
+        return self.negate ^ matches(inferred, nodes.Const(None))
```

Comment:
```suggestion
        return self.negate ^ matches(inferred, self.CONST_NONE)
```
Or `NoneConstraint.CONST_NONE`